cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -442,43 +452,45 @@ class device_row_comparator {
       // and empty lists, so not every index corresponds to an actual element
       // in the child column. The element_index is used to keep track of the current
       // child element that we're actually comparing.
-      weak_ordering state{weak_ordering::EQUIVALENT};
       for (int l_dremel_index = l_start, r_dremel_index = r_start, element_index = 0;
            l_dremel_index < l_end and r_dremel_index < r_end;
            ++l_dremel_index, ++r_dremel_index) {
-        // First early exit: the definition levels do not match.
-        if (l_def_levels[l_dremel_index] != r_def_levels[r_dremel_index]) {
-          state = (l_def_levels[l_dremel_index] < r_def_levels[r_dremel_index])
-                    ? weak_ordering::LESS
-                    : weak_ordering::GREATER;
-          return cuda::std::pair(state, _depth);
-        }
-
-        // Second early exit: the repetition levels do not match.
-        if (l_rep_levels[l_dremel_index] != r_rep_levels[r_dremel_index]) {
-          state = (l_rep_levels[l_dremel_index] < r_rep_levels[r_dremel_index])
-                    ? weak_ordering::LESS
-                    : weak_ordering::GREATER;
-          return cuda::std::pair(state, _depth);
-        }
+        auto const l_rep_level = l_rep_levels[l_dremel_index];
+        auto const r_rep_level = r_rep_levels[r_dremel_index];
+
+        // only compare if left and right are at same nesting level
+        if (l_rep_level == r_rep_level) {
```

Comment:
Can you write this as an "early exit" so that the `else` block isn't hidden far below? It should also help reduce an indentation level for the body of this `if`.